servers/src/main/java/tachyon/worker/lineage/LineageDataManager.java

```diff
@@ -0,0 +1,157 @@
+    // Create Under FileSystem Client
+    String ufsAddress = mTachyonConf.get(Constants.UNDERFS_ADDRESS);
+    mUfs = UnderFileSystem.get(ufsAddress, mTachyonConf);
+  }
+
+  /**
+   * Persists the blocks of a file into the under file system.
+   *
+   * @param fileId the id of the file.
+   * @param blockIds the list of block ids.
+   * @throws IOException
+   */
+  public synchronized void persistFile(long fileId, List<Long> blockIds)
+      throws IOException {
+    String ufsDataFolder = mTachyonConf.get(Constants.UNDERFS_DATA_FOLDER);
+    FileInfo fileInfo;
+    try {
+      fileInfo = mBlockDataManager.getFileInfo(fileId);
+    } catch (FileDoesNotExistException e) {
+      LOG.warn("Fail to get file info for file " + fileId, e);
+      throw new IOException(e);
+    }
+    TachyonURI uri = new TachyonURI(fileInfo.getPath());
+    String dstPath = PathUtils.concatPath(ufsDataFolder, fileInfo.getPath());
+    LOG.info("persist file " + fileId + " at " + dstPath);
+    String parentPath = PathUtils.concatPath(ufsDataFolder, uri.getParent().getPath());
+    if (!mUfs.exists(parentPath) && !mUfs.mkdirs(parentPath, true)) {
+      throw new IOException("Failed to create " + parentPath);
+    }
+    OutputStream outputStream = mUfs.create(dstPath);
+    final WritableByteChannel outputChannel = Channels.newChannel(outputStream);
+
+    for (long blockId : Lists.newArrayList(blockIds)) {
+      long lockId;
+      try {
+        lockId = mBlockDataManager.lockBlock(Sessions.CHECKPOINT_SESSION_ID, blockId);
+      } catch (NotFoundException ioe) {
+        throw new IOException(ioe);
+      }
+
+      // obtain block reader
+      try {
+        BlockReader reader;
+        try {
+          reader =
+              mBlockDataManager.readBlockRemote(Sessions.CHECKPOINT_SESSION_ID, blockId, lockId);
+        } catch (NotFoundException nfe) {
+          throw new IOException(nfe);
+        } catch (InvalidStateException ise) {
+          throw new IOException(ise);
+        }
+
+        // write content out
+        ReadableByteChannel inputChannel = reader.getChannel();
+        fastCopy(inputChannel, outputChannel);
+        reader.close();
+      } finally {
+        try {
+          mBlockDataManager.unlockBlock(lockId);
+        } catch (NotFoundException nfe) {
+          throw new IOException(nfe);
+        }
+      }
+    }
+
+    outputStream.flush();
+    outputChannel.close();
+    outputStream.close();
+    mPersistedFiles.add(fileId);
+  }
+
+
+  private static void fastCopy(final ReadableByteChannel src, final WritableByteChannel dest)
+      throws IOException {
+    final ByteBuffer buffer = ByteBuffer.allocateDirect(16 * 1024);
+
+    while (src.read(buffer) != -1) {
+      buffer.flip();
+      dest.write(buffer);
+      buffer.compact();
+    }
+
+    buffer.flip();
+
+    while (buffer.hasRemaining()) {
+      dest.write(buffer);
+    }
+  }
+
+  public synchronized List<Long> fetchPersistedFiles() {
```

Comment:
Because usually a getter doesn't change the state. Perhaps `popPersistedFiles` sounds better?